### PR TITLE
CVE-2019-1019: Bypass SMB singing for unpatched machines

### PR DESCRIPTION
**Pull request by msimakov**
https://portal.msrc.microsoft.com/en-US/security-guidance/advisory/CVE-2019-1019

Bypassing SMB singing by removing the target in NTLMSSP_CHALLENGE messages

### CODE REVIEW

**Comment by asolino** [comment]
Hey @msimakov 

Thanks for the PR.. Was very much looking forward for you to add your discoveries ;). Clever addition to the previous one (CVE-2015-0005), congrats.

I'll be testing it today. Windows SMB Clients still not working? (throwing INVALID_PARAMETER on your side?)

**Comment by msimakov** [comment]
Hi @asolino 

Thanks, glad to contribute ;) Let me know if you run into any issues.
Windows SMB Clients still not working, they respond with a TCP RST and re-negotiate the SMB connection.

**Comment by asolino** [comment]
Having some problems making it work @msimakov 
Running it this way:
```
ntlmrelayx.py -machine-account FREEFLY.NET/machine\$ -machine-hashes LM:NT -domain FREEFLY.NET -t target -debug -remove-target -smb2support
```
I'm adding some comments in your code

File: impacket/examples/ntlmrelayx/clients/smbrelayclient.py [hunk 1]
```diff
@@ -350,10 +473,35 @@ def sendAuth(self, authenticateMessageBlob, serverChallenge=None):
...
+            respToken2 = SPNEGO_NegTokenResp(authData)
+            authenticateMessageBlob = respToken2['ResponseToken']
+
+            errorCode, signingKey = self.netlogonSessionKey(authData)
+
+            # Recalculate MIC
+            res = NTLMAuthChallengeResponse()
+            res.fromString(authenticateMessageBlob)
+
+            new_auth_blob = hexlify(authenticateMessageBlob)[0:144] + b'00000000000000000000000000000000' + hexlify(authenticateMessageBlob)[176:]
```
Comment (asolino) [inline impacket/examples/ntlmrelayx/clients/smbrelayclient.py]: This just doesn't sound right. Are you trying to clear the previous MIC in order to calculate the HMAC_MD5 or something else?

Comment (msimakov) [reply impacket/examples/ntlmrelayx/clients/smbrelayclient.py]: I'm clearing the previous MIC in order to calculate the new one (take a look at the next line - the relay_MIC)

**Comment by asolino** [comment]
Merged! Thanks a lot @msimakov !

File: impacket/examples/ntlmrelayx/clients/smbrelayclient.py (hunk 1 above)
Comment (zur250) [reply impacket/examples/ntlmrelayx/clients/smbrelayclient.py]: hi @msimakov , i am trying to use the script by myself and i am getting some errors. Basically i can see that i am able to retrieve the session key, but for some reason i cannot connect, i am using the following command just like @asolino said to use it in order to perform some tests.
by looking at wireshark on the relay endpoint i can see that, everything goes just like the article you published, but at the last message ( the one wit the new mic) i am getting Error:STATUS_INVALID_PARAMETER and the final login is failing.
Any ideas?

Comment (msimakov) [reply impacket/examples/ntlmrelayx/clients/smbrelayclient.py]: Hi @zur250. Make sure to run the updated version, the file has been updated after this commit

Comment (asolino) [reply impacket/examples/ntlmrelayx/clients/smbrelayclient.py]: Hey @zur250. I think you have a mix of impacket versions installed. What is the banner you get when you run `ntlmrelayx.py`? (at the very beginning of its execution)

**Pull request merged**